plugins/org.python.pydev/src/org/python/pydev/editor/correctionassist/heuristics/AssistFString.java

```diff
@@ -0,0 +1,181 @@
+import org.eclipse.jface.text.ITypedRegion;
+import org.python.pydev.core.IPyEdit;
+import org.python.pydev.core.IPythonNature;
+import org.python.pydev.core.MisconfigurationException;
+import org.python.pydev.core.docutils.ParsingUtils;
+import org.python.pydev.core.docutils.PySelection;
+import org.python.pydev.core.docutils.SyntaxErrorException;
+import org.python.pydev.core.log.Log;
+import org.python.pydev.core.partition.PyPartitionScanner;
+import org.python.pydev.core.proposals.CompletionProposalFactory;
+import org.python.pydev.editor.PyEdit;
+import org.python.pydev.editor.codefolding.PySourceViewer;
+import org.python.pydev.editor.correctionassist.IAssistProps;
+import org.python.pydev.shared_core.code_completion.ICompletionProposalHandle;
+import org.python.pydev.shared_core.code_completion.IPyCompletionProposal;
+import org.python.pydev.shared_core.image.IImageCache;
+import org.python.pydev.shared_core.image.IImageHandle;
+import org.python.pydev.shared_core.image.UIConstants;
+import org.python.pydev.shared_core.partitioner.FastPartitioner;
+import org.python.pydev.shared_core.string.FastStringBuffer;
+import org.python.pydev.shared_core.string.StringUtils;
+
+public class AssistFString implements IAssistProps {
+    @Override
+    public List<ICompletionProposalHandle> getProps(PySelection ps, IImageCache imageCache, File f,
+            IPythonNature nature, IPyEdit edit, int offset) throws BadLocationException, MisconfigurationException {
+        PySourceViewer viewer = null;
+        if (edit instanceof PyEdit) { //only in tests it is actually null
+            viewer = ((PyEdit) edit).getPySourceViewer();
+        }
+        List<ICompletionProposalHandle> lst = new ArrayList<ICompletionProposalHandle>();
+
+        IDocument doc = ps.getDoc();
+        ITypedRegion partition = ((FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc)).getPartition(offset);
+        if (!ParsingUtils.isStringContentType(partition.getType())) {
+            return lst;
+        }
+
+        final String stringPartitionContents = doc.get(partition.getOffset(), partition.getLength());
+        int formatCount = StringUtils.count(stringPartitionContents, "%s");
+        if (formatCount == 0) {
+            return lst;
+        }
+
+        // get string end offset
+        int partitionEndOffset = partition.getOffset() + partition.getLength();
+        // get format end offset
+        int i = partitionEndOffset;
+        try {
+            ParsingUtils parsingUtils = ParsingUtils.create(doc);
+            while (i < parsingUtils.len()) {
+                char c = parsingUtils.charAt(i);
+                if (Character.isWhitespace(c)) {
+                    i++;
+                    continue;
+                }
+
+                if (c == '{') {
+                    return lst;
+                }
+
+                if (c == '[' || c == '(' || c == '"' || c == '\'') {
+                    i = parsingUtils.eatPar(i, null, c) + 1;
```

Comment:
We should check for cases where the code is not well formed here.

i.e.: 

`'%s' % (somevar`

-- note that the closing `)` is not there.

(no suggestions are needed in this case).

-- I think this already works, but we should add the test